### PR TITLE
[TVMSCRIPT] TVMScript Parser support BufferSlice indices

### PR DESCRIPTION
**Pull request by Beya2019**
TVMScript Parser not support buffer as indices, such as index_buf[0] is the indices of data_buf[vi, vj, index_buf[0]]
```
@tvm.script.tir
def func(data: ty.handle, index: ty.handle) -> None:
    data_buf = tir.match_buffer(data, (1, 5, 6), "float16")
    index_buf = tir.match_buffer(index, (1,), "int32")
    copy_buf = tir.alloc_buffer((1, 5), "float16")

    with tir.block([1, 5], "init") as [vi, vj]:
        copy_buf[vi, vj] = data_buf[vi, vj, index_buf[0]]

```
we add the code to support this function, so we can get the result when call `tvm.script.asscript(func)`
```
primfn(data: handle, index: handle) -> ()
  buffers = {index_buf: Buffer(index_buf_1: Pointer(int32), int32, [1], []),
             data_buf: Buffer(data_buf_1: Pointer(float16), float16, [1, 5, 6], [])}
  buffer_map = {data: data_buf, index: index_buf} {
  block([], "root") {
    tir.reads([])
    tir.writes([])
    copy_buf = alloc_buffer(float16[1, 5])
    for (i0: int32, 0, 1) {
      for (i1: int32, 0, 5) {
        block([1, 5], "init") as [vi, vj] {
          bind(vi, i0)
          bind(vj, i1)
          tir.reads([data_buf[vi, vj, 0:6], index_buf[0]])
          tir.writes([copy_buf[vi, vj]])
          copy_buf[vi, vj] = data_buf[vi, vj, index_buf[0]]
      }
    }
}

```
@Hzfengsy Would you please have a look at this? Thanks very much.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @Hzfengsy please help to take a look

File: python/tvm/script/node.py [hunk 1]
```diff
@@ -114,6 +114,11 @@ def check_index(index: Union[int, PrimExpr]):
                 check_index(index.start)
                 check_index(index.stop)
                 slices.append(index)
+            elif isinstance(index, BufferSlice):
+                for s in index.slices:
+                    if isinstance(s, Slice):
+                        check_index(s.start)
+                slices.append(index)
```
Comment (tkonolige) [inline python/tvm/script/node.py:121]: I don't think this is correct. I believe you should should assume that the BufferSlice already checked that its indices were correct. Instead you want to check if the buffer dtype is correct (int32).

I am also a little confused as to why this case is not handled by the PrimExpr case?

File: python/tvm/script/node.py [hunk 2]
```diff
@@ -133,18 +138,27 @@ def check_index(index: Union[int, PrimExpr]):
...
-                regions.append(str(s.start))
-            else:
-                regions.append(str(s.start) + ": " + str(s.stop))
+            if isinstance(s, Slice):
+                if s.stop is None:
+                    regions.append(str(s.start))
+                else:
+                    regions.append(str(s.start) + ": " + str(s.stop))
+            elif isinstance(s, BufferSlice):
+                regions.append(s.buffer.name + "[" + str(s.start) + "]")
```
Comment (tkonolige) [inline python/tvm/script/node.py:147]: You want this to recurse, so just call str on the BufferSlice.

File: python/tvm/script/node.py [hunk 3]
```diff
@@ -133,18 +138,27 @@ def check_index(index: Union[int, PrimExpr]):
...
-
-        indices = [s.start for s in self.slices]
+            if isinstance(s, Slice):
+                if s.stop is not None:
+                    self.report_error("BufferLoad only accepts elementwise access", self.span)
+                indices.append(s.start)
+            elif isinstance(s, BufferSlice):
+                args: List[PrimExpr] = []
+                for idx in s.slices:
+                    args.append(idx.start)
```
Comment (tkonolige) [inline python/tvm/script/node.py:162]: Why is the conversion from from BufferSlice to BufferLoad necessary here? Plus it seems like this should be handled by the recursive case.

File: tests/python/unittest/test_tvmscript_ops.py [hunk 4]
```diff
@@ -103,5 +103,46 @@ def test_get_valid_counts_script_func():
...
 
 
+@tvm.script.tir
+def ops_with_buffer_slice_indices(data: ty.handle, index: ty.handle) -> None:
+    data_buf = tir.match_buffer(data, (1, 5, 6), "float16")
+    index_buf = tir.match_buffer(index, (1,), "int32")
+    copy_buf = tir.alloc_buffer((1, 5), "float16")
+
+    with tir.block([1, 5], "init") as [vi, vj]:
+        copy_buf[vi, vj] = data_buf[vi, vj, index_buf[0]]
```
Comment (tkonolige) [inline tests/python/unittest/test_tvmscript_ops.py:113]: Can you add the following tests:
- a test that nests buffer accesses three deep. i.e. `data_ buf[index_buf[index_buf[0]]]]`
- a test where the inner buffer dtype is not int32 (should fail).
- a test where the inner buffer is a slice. i.e. `data_buf[index_buf[0,1:2]]`

**Review by tkonolige — CHANGES_REQUESTED** [review]

**Comment by Hzfengsy** [comment]
I recommend converting BufferSlice to BufferLoad when necessary. It can help:
1. recursively BufferLoad
2. Index Type check
3. Printing

File: python/tvm/script/node.py (hunk 1 above)
Comment (Hzfengsy) [reply python/tvm/script/node.py:121]: Python does not have type conversion system. The only thing we do is converting `BufferSlice` to `BufferLoad` when creating TVM Object. 

In this case, it still in the Python system, so can not be converted into `BufferLoad`

File: tests/python/unittest/test_tvmscript_ops.py (hunk 4 above)
Comment (Beya2019) [reply tests/python/unittest/test_tvmscript_ops.py:113]: Follow the suggestion of @Hzfengsy, I convert the BufferSlice to BufferLoad in node.py

Meanwhile, add the 3 test cases listed above
1. add `func_with_bufferslice_indices` and `func_with_recursive_bufferslice_indices` in test_tvmscript_complete.py
2. add the fail test case error_bufferslice_index_type in test_tvmscript_error_report.py
3. Given that BufferLoad only accepts elementwise access, add the fail case `error_index_with_stop` `error_bufferslice_index_with_stop` in test_tvmscript_error_report.py

**Review by Hzfengsy — CHANGES_REQUESTED** [review]

File: tests/python/unittest/test_tvmscript_complete.py [hunk 5]
```diff
@@ -190,9 +190,77 @@ def render(e):
...
+    index_buf = tir.match_buffer(index, (1,), "int32")
+    out_buf = tir.alloc_buffer((16, 16), "float32")
+
+    with tir.block([16, 16]) as [vi, vj]:
+        out_buf[vi, vj] = data_buf[vi, index_buf[0]]
+
+
+@tvm.script.tir
+def expected_bufferslice_indices(data: ty.handle, index: ty.handle) -> None:
+    index_buf = tir.match_buffer(
```
[the remark this answers is not in the document]
Comment (Beya2019) [reply tests/python/unittest/test_tvmscript_complete.py:205]: OK, Get it. Thanks for reminding.

**Review by tkonolige — APPROVED** [review]
Looks good to me. Thank you!

**Review by Hzfengsy — APPROVED** [review]
LGTM, Thanks for the contribution!

**Review by tqchen — APPROVED** [review]

**Comment by tqchen** [comment]
Thanks @Beya2019 @Hzfengsy @tkonolige !

**Pull request merged**